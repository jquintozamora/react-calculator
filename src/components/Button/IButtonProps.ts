export interface IButtonProps {
  "data-name": string;
  onClick?: (event: IButtonMouseEvent) => void;
  expand?: boolean;
  color?: string;
  bgColor?: string;
  borderColor?: string;
  hoverColor?: string
  selected?: boolean;
}

export interface IButtonContainerProps {
  expand?: boolean;
}

interface IButtonEventTarget extends EventTarget {
  dataset?: {
    name: string;
  };
}

export interface IButtonMouseEvent extends React.MouseEvent<HTMLElement> {
  target: IButtonEventTarget;
}
