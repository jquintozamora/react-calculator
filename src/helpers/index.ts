import { ICalculatorState } from "../components/Calculator/ICalculatorState";

export function makeOperation(total: number, nextNumber: number, operator: string) {
  switch (operator) {
    case "+":
      return total + nextNumber;
    case "-":
      return total - nextNumber;
    case "x":
      return total * nextNumber;
    case "รท":
      return total / nextNumber;
    default:
      return null;
  }
}

export function calculateState(state: ICalculatorState, name: string): ICalculatorState {
  const { total, nextNumber, operator } = state;

  if (name === "AC") {
    return { total: null, nextNumber: null, operator: null };
  }

  if (name === "+/-") {
    if (nextNumber !== null) {
      return { ...state, nextNumber: (parseFloat(nextNumber) * -1).toString() };
    }
    if (total !== null) {
      return { ...state, total: (parseFloat(total) * -1).toString() };
    }
  }

  if (name === ".") {
    if (total === null) {
      return { ...state, total: "0." };
    }
    if (total !== null && operator === null) {
      if (total.includes(".") === false) {
        return { ...state, total: total + "." };
      }
    }
    if (total !== null && operator !== null) {
      if (nextNumber === null) {
        return { ...state, nextNumber: "0." };
      } else {
        if (nextNumber.includes(".") === false) {
          return { ...state, nextNumber: nextNumber + "." };
        }
      }
    }
  }

  if (name === "%") {
    if (nextNumber !== null) {
      return { ...state, nextNumber: (parseFloat(nextNumber) / 100).toString() };
    }
    if (total !== null) {
      return { ...state, total: (parseFloat(total) / 100).toString() };
    }
  }

  if (name === "=" && total !== null && operator !== null && nextNumber !== null) {
    const result = makeOperation(parseFloat(total), parseFloat(nextNumber), operator);
    if (result !== null) {
      return { total: result.toString(), nextNumber: null, operator: null };
    }
  }

  const operators = ["+", "-", "x", "รท"];
  const operatorNext = operators.includes(name) && name;
  if (operatorNext) {
    if (total !== null) {
      return { ...state, operator: operatorNext };
    }
  }

  const numbers = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"];
  const number = numbers.includes(name) && name;
  if (number !== false) {
    // first time user input a number
    if (total === null) {
      return { ...state, total: number };
    }

    // user continue filling first operator
    if (total !== null && operator === null) {
      const concatTotal = parseFloat(total.toString() + number);
      return { ...state, total: concatTotal.toString() };
    }

    // we already got first operator and operation set up
    if (total !== null && operator !== null) {
      if (nextNumber === null) {
        return { ...state, nextNumber: number };
      } else {
        const concatNextNumber = parseFloat(nextNumber.toString() + number);
        return { ...state, nextNumber: concatNextNumber.toString() };
      }
    }
  }

  return state;
}
