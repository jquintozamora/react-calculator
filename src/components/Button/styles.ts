import styled from "styled-components";
import { ifProp, prop } from "styled-tools";
import { IButtonContainerProps, IButtonProps } from "./IButtonProps";

export const ButtonContainer = styled.div<IButtonContainerProps>`
  flex: ${ifProp({ expand: true }, "2", "1")};
`;

export const RoundedButton = styled.div<IButtonProps>`
  background-color: ${prop("bgColor")};
  color: ${prop("color")};
  cursor: pointer;
  font-size: 3rem;
  font-weight: bold;
  text-align: center;
  border: 1px solid ${prop("borderColor")};
  margin: 5px;
  line-height: 50px;
  user-select: none;
  opacity: ${ifProp({ selected: true }, "0.5", "1")};

  &:hover {
    background-color: ${prop("borderColor")};
    color: ${prop("hoverColor")};
  }
`;
