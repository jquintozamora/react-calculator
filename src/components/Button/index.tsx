import * as React from "react";
import { eeBlue, grey, lightGrey, black } from "../../styles/colors";
import { IButtonProps } from "./IButtonProps";
import { ButtonContainer, RoundedButton } from "./styles";

export const Button = (props: IButtonProps) => {
  const { expand, ...rest } = props;
  return (
    <ButtonContainer expand={expand}>
      <RoundedButton {...rest}>{props["data-name"]}</RoundedButton>
    </ButtonContainer>
  );
};

Button.defaultProps = {
  bgColor: lightGrey,
  color: eeBlue,
  borderColor: grey,
  hoverColor: black
};

export default React.memo(Button);
