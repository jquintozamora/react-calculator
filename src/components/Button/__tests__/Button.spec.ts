import Button from "..";
import { buildTestShallowRender, buildTestRender } from "../../../../test/unit/helpers/ComponentHelper";
import {
  eeBlue,
  eeLightBlue,
  eeDarkYellow,
  eeYellow,
  eeDarkGrey,
  grey,
  lightGrey,
  lightBlack,
  white,
  black
} from "../../../styles/colors";

describe("<Button />", () => {
  const renderComponent = buildTestShallowRender(Button);
  const defaultProps = {
    "data-name": "1",
    onClick: jest.fn(),
    expand: false,
    bgColor: lightGrey,
    color: eeBlue,
    borderColor: grey,
    hoverColor: grey,
    selected: false
  };

  describe("@renders", () => {
    it("should render default state", () => {
      expect(renderComponent().tree).toMatchSnapshot();
    });

    it("should render numeric button", () => {
      expect(
        renderComponent({ ...defaultProps, "data-name": "2", bgColor: lightBlack, color: white }).tree
      ).toMatchSnapshot();
    });

    it("should render operation button", () => {
      expect(
        renderComponent({
          ...defaultProps,
          "data-name": "+",
          bgColor: eeYellow,
          color: black,
          borderColor: eeDarkYellow,
          selected: false
        }).tree
      ).toMatchSnapshot();
    });

    it("should render operation button selected", () => {
      expect(
        renderComponent({
          ...defaultProps,
          "data-name": "-",
          bgColor: eeYellow,
          color: black,
          borderColor: eeDarkYellow,
          selected: true
        }).tree
      ).toMatchSnapshot();
    });

    it("should render others button", () => {
      expect(
        renderComponent({
          ...defaultProps,
          "data-name": "AC",
          bgColor: eeLightBlue,
          borderColor: eeBlue,
          color: eeDarkGrey,
          hoverColor: white
        }).tree
      ).toMatchSnapshot();
    });
  });

  describe("@events", () => {
    const renderComponentEv = buildTestRender(Button);
    it("should call onClick function", () => {
      const mockOnclick = jest.fn();
      const { root } = renderComponentEv({ ...defaultProps, onClick: mockOnclick });
      expect(mockOnclick).toHaveBeenCalledTimes(0);
      const button = root.find(el => el.props["data-name"] === "1");
      button.props.onClick();
      expect(mockOnclick).toHaveBeenCalledTimes(1);
    });
  });
});
