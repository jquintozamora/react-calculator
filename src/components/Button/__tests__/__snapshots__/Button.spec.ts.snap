// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> @renders should render default state 1`] = `
.c0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  background-color: #f5f5f5;
  color: #1795d4;
  cursor: pointer;
  font-size: 3rem;
  font-weight: bold;
  text-align: center;
  border: 1px solid #949494;
  margin: 5px;
  line-height: 50px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  opacity: 1;
}

.c1:hover {
  background-color: #949494;
  color: #000000;
}

<div
  className="c0"
>
  <div
    className="c1"
    color="#1795d4"
  />
</div>
`;

exports[`<Button /> @renders should render numeric button 1`] = `
.c0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  background-color: #373d45;
  color: #FFFFFF;
  cursor: pointer;
  font-size: 3rem;
  font-weight: bold;
  text-align: center;
  border: 1px solid #949494;
  margin: 5px;
  line-height: 50px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  opacity: 1;
}

.c1:hover {
  background-color: #949494;
  color: #949494;
}

<div
  className="c0"
>
  <div
    className="c1"
    color="#FFFFFF"
    data-name="2"
    onClick={[MockFunction]}
    selected={false}
  >
    2
  </div>
</div>
`;

exports[`<Button /> @renders should render operation button 1`] = `
.c0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  background-color: #fed800;
  color: #000000;
  cursor: pointer;
  font-size: 3rem;
  font-weight: bold;
  text-align: center;
  border: 1px solid #d4b526;
  margin: 5px;
  line-height: 50px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  opacity: 1;
}

.c1:hover {
  background-color: #d4b526;
  color: #949494;
}

<div
  className="c0"
>
  <div
    className="c1"
    color="#000000"
    data-name="+"
    onClick={[MockFunction]}
    selected={false}
  >
    +
  </div>
</div>
`;

exports[`<Button /> @renders should render operation button selected 1`] = `
.c0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  background-color: #fed800;
  color: #000000;
  cursor: pointer;
  font-size: 3rem;
  font-weight: bold;
  text-align: center;
  border: 1px solid #d4b526;
  margin: 5px;
  line-height: 50px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  opacity: 0.5;
}

.c1:hover {
  background-color: #d4b526;
  color: #949494;
}

<div
  className="c0"
>
  <div
    className="c1"
    color="#000000"
    data-name="-"
    onClick={[MockFunction]}
    selected={true}
  >
    -
  </div>
</div>
`;

exports[`<Button /> @renders should render others button 1`] = `
.c0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  background-color: #cdeeff;
  color: #666;
  cursor: pointer;
  font-size: 3rem;
  font-weight: bold;
  text-align: center;
  border: 1px solid #1795d4;
  margin: 5px;
  line-height: 50px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  opacity: 1;
}

.c1:hover {
  background-color: #1795d4;
  color: #FFFFFF;
}

<div
  className="c0"
>
  <div
    className="c1"
    color="#666"
    data-name="AC"
    onClick={[MockFunction]}
    selected={false}
  >
    AC
  </div>
</div>
`;
